File: Problems/19_Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional
from Operation_of_ListNode import *

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        slow = head
        slow_pre = head
        while fast:
            if n<=0:
                slow_pre = slow
                slow = slow.next
            fast = fast.next
            n -= 1
        slow_pre.next = slow.next
        # Slow is about to move
        if n==0:
            return head.next
        # Didn't even move
        if slow==slow_pre:
            return None
        # regular case
        return head

def main():
    sol = Solution()
    head = List2Node([1,2])
    n = 1
    print(Node2List(sol.removeNthFromEnd(head, n)))
if __name__ == '__main__':
    main()

File: Problems/19_Remove_Nth_Node_From_End_of_List.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // dummyHead
        ListNode* ans = new ListNode(0);
        ans->next = head;
        ListNode* slow = ans;
        ListNode* fast = ans;
        for(int i=0; i<n; i++){
            fast = fast->next;
        }
        cout<<fast->val;
        while(fast->next!=NULL){
            fast = fast->next;
            slow = slow->next;
        }
        slow->next = slow->next->next;
        return ans->next;
    }
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* fast = head;
        ListNode* slow = head;
        ListNode* slow_pre = head;
        while(fast != nullptr){
            if(n <= 0){
                slow_pre = slow;
                slow = slow->next;
            }
            fast = fast->next;
            n--;
        }
        slow_pre->next = slow->next;
        if(n==0){
            return head->next;
        }
        if(slow_pre==slow){
            return nullptr;
        }
        return head;
    }
};